src/stash/src/upgrade/json_to_proto.rs

```diff
@@ -0,0 +1,1872 @@
+    collection_id: i64,
+    epoch: NonZeroI64,
+) -> Result<(), anyhow::Error>
+where
+    KJ: for<'de> Deserialize<'de>,
+    VJ: for<'de> Deserialize<'de>,
+    KP: ::prost::Message + From<KJ>,
+    VP: ::prost::Message + From<VJ>,
+{
+    // Get all of the rows for a single collection and convert them to proto.
+    let mut proto_rows = client
+        .query(
+            "SELECT key, value, time, diff FROM data WHERE collection_id = $1",
+            &[&collection_id],
+        )
+        .await
+        .context("getting rows")?
+        .into_iter()
+        .map(|row| {
+            let key: serde_json::Value = row.get("key");
+            let val: serde_json::Value = row.get("value");
+            let time: i64 = row.get("time");
+            let diff: i64 = row.get("diff");
+
+            let key_json: KJ = serde_json::from_value(key).expect("must deserialize");
+            let val_json: VJ = serde_json::from_value(val).expect("must deserialize");
+
+            let key_proto = KP::from(key_json);
+            let val_proto = VP::from(val_json);
+
+            ((key_proto, val_proto), time, diff)
+        });
+
+    // Insert all of the converted protos into the new table.
+    loop {
+        let mut batch = Vec::new();
+        let mut rows = 0;
+        let mut serialized_size = 0;
+
+        // Keep popping rows into the batch.
+        while let Some(((key, val), time, diff)) = proto_rows.next() {
+            serialized_size += key.encoded_len();
+            serialized_size += val.encoded_len();
+
+            batch.push(((key.encode_to_vec(), val.encode_to_vec()), time, diff));
+            rows += 1;
+
+            // Once our serialized size is greater than our max, break to run an insert.
+            if serialized_size > INSERT_BATCH_MAX_SIZE {
+                break;
+            }
+        }
+
+        // We must have removed all of the rows, exit!
+        if rows == 0 {
+            return Ok(());
+        }
+
+        // Prepare our insert statement.
+        let mut stmt =
+            String::from("INSERT INTO data_proto (collection_id, key, value, time, diff) VALUES");
+        let mut sep = ' ';
+        for i in 0..rows {
+            let idx = 1 + i * 4;
+            write!(
+                &mut stmt,
+                "{}($1, ${}, ${}, ${}, ${})",
+                sep,
+                idx + 1,
+                idx + 2,
+                idx + 3,
+                idx + 4
+            )
+            .unwrap();
+            sep = ',';
+        }
+        let stmt = client.prepare(&stmt).await.context("preparing insert")?;
+
+        // Prepare our args.
+        let mut args: Vec<&'_ (dyn ToSql + Sync)> = vec![&collection_id];
+        for ((key, val), time, diff) in &batch {
+            args.push(key);
+            args.push(val);
+            args.push(time);
+            args.push(diff);
+        }
+
+        // Validate our epoch hasn't changed.
+        check_fence(|stmt| client.query_one(stmt, &[]), epoch)
```

Comment:
Gotta do this check in the same sql transaction as the insert or it's not safe.

Reply:
Done!